monitoring: alert admins when critical alerts are firing

bobheadxi opened the pull request:
Stacked PR on top of #11832. Adds a site alert (for admins only) when critical alerts are firing

Closes #11517

<img width="545" alt="image" src="https://user-images.githubusercontent.com/23356519/86601425-e0dff600-bfd3-11ea-9a9c-9e1184a98f8a.png">


<!-- Reminder: Have you updated the changelog and relevant docs (user docs, architecture diagram, etc) ? -->

slimsag commented on cmd/frontend/graphqlbackend/site_alerts.go:
```diff
@@ -264,3 +271,50 @@ func determineOutOfDateAlert(isAdmin bool, months int, offline bool) *Alert {
+		errorKey := "active-alerts-alert-error"
+		promURL, err := url.Parse(prometheusURL)
+		if err != nil {
+			return []*Alert{{TypeValue: AlertTypeWarning, MessageValue: fmt.Sprintf("Prometheus misconfigured: %s", err), IsDismissibleWithKeyValue: errorKey}}
```
Due to the dismissible key, if someone dismisses this error once and it occurs again they would not get to see it again

slimsag commented on cmd/frontend/graphqlbackend/site_alerts.go:
```diff
@@ -264,3 +271,50 @@ func determineOutOfDateAlert(isAdmin bool, months int, offline bool) *Alert {
+		promURL.Path = "/prom-wrapper/alerts-status"
+		req, err := http.NewRequest("GET", promURL.String(), nil)
+		if err != nil {
+			return []*Alert{{TypeValue: AlertTypeWarning, MessageValue: fmt.Sprintf("Prometheus misconfigured: %s", err), IsDismissibleWithKeyValue: errorKey}}
```
same here

slimsag commented on cmd/frontend/graphqlbackend/site_alerts.go:
```diff
@@ -264,3 +271,50 @@ func determineOutOfDateAlert(isAdmin bool, months int, offline bool) *Alert {
+		defer cancel()
+		resp, err := http.DefaultClient.Do(req.WithContext(ctx))
+		if err != nil {
+			return []*Alert{{TypeValue: AlertTypeWarning, MessageValue: fmt.Sprintf("Unable to fetch alerts status: %s", err), IsDismissibleWithKeyValue: errorKey}}
```
and here

slimsag commented on cmd/frontend/graphqlbackend/site_alerts.go:
```diff
@@ -264,3 +271,50 @@ func determineOutOfDateAlert(isAdmin bool, months int, offline bool) *Alert {
+			return []*Alert{{TypeValue: AlertTypeWarning, MessageValue: fmt.Sprintf("Unable to fetch alerts status: %s", err), IsDismissibleWithKeyValue: errorKey}}
+		}
+		if resp.StatusCode != 200 {
+			return []*Alert{{TypeValue: AlertTypeWarning, MessageValue: fmt.Sprintf("Unable to fetch alerts status: status %d", resp.StatusCode), IsDismissibleWithKeyValue: errorKey}}
```
and here

slimsag commented on cmd/frontend/graphqlbackend/site_alerts.go:
```diff
@@ -264,3 +271,50 @@ func determineOutOfDateAlert(isAdmin bool, months int, offline bool) *Alert {
+		var alertsStatus map[string]int
+		defer resp.Body.Close()
+		if err := json.NewDecoder(resp.Body).Decode(&alertsStatus); err != nil {
+			return []*Alert{{TypeValue: AlertTypeWarning, MessageValue: err.Error(), IsDismissibleWithKeyValue: errorKey}}
```
and here

slimsag commented on cmd/frontend/graphqlbackend/site_alerts.go:
```diff
@@ -264,3 +271,50 @@ func determineOutOfDateAlert(isAdmin bool, months int, offline bool) *Alert {
+		if criticalAlerts == 0 {
+			return nil
+		}
+		msg := fmt.Sprintf("%d critical alerts across %d services are currently firing - [view alerts](/-/debug/grafana)",
```
Can we adjust the wording to reflect the case of 1 critical alert (non-plural) and 1 service (non-plural)?

bobheadxi replied:
f74b1dd

slimsag commented:
This looks awesome!

The biggest worry I have here is just that the way we have dismissible keys set up https://github.com/sourcegraph/sourcegraph/pull/11957#discussion_r451275824 means even if you dismiss a single alert - it's never going to reset - you'll never see this notification again which is quite surprising.

It likely wouldn't be easy to solve this because we don't have good unique keys in this case, so I would propose the following easy solution:

1. All notifications added in this PR are non-dismissible.
2. If `"observability.hideNotifications": true,` is set in the site config, all site alert notifications will be hidden (if critical alerts are firing, if there are Prometheus misconfigurations, etc.)

Thoughts?

bobheadxi commented:
> If "observability.hideNotifications": true, is set in the site config, all site alert notifications will be hidden (if critical alerts are firing, if there are Prometheus misconfigurations, etc.)

Would Prometheus misconfigurations include Alertmanager (notifier) configuration?

slimsag commented:
> Would Prometheus misconfigurations include Alertmanager (notifier) configuration?

Yeah, I think so - my thinking here is that we should aim for the ideal case being nobody ever has to turn these off and therefor them being non-dismissible is not a problem. But, in the event that we've missed something - these are particularly noisy - or contacting prometheus/alertmanager is failing for some reason and a customer needs a subsequent update from us - we can just instruct them to hide these notifications until we fix the issue.

Maybe `observability.debugHideSiteAlerts` would be a better name?

bobheadxi commented:
@stephen I've added `user.hideObservabilitySiteAlerts` in b556195399:

* top-level problems are reported by `observabilityActiveAlertsAlert()`, can only access viewer settings here, so I added `hideObservabilitySiteAlerts` to viewer config
* this conf is not accessibly for the conf validator, so the escape hatch for `observability.alerts` problems is to just remove `observability.alerts`

slimsag commented on cmd/frontend/graphqlbackend/site_alerts.go:
```diff
@@ -264,3 +271,57 @@ func determineOutOfDateAlert(isAdmin bool, months int, offline bool) *Alert {
+// observabilityActiveAlertsAlert directs admins to check Grafana if critical alerts are firing
+func observabilityActiveAlertsAlert(prometheusURL string) func(AlertFuncArgs) []*Alert {
+	return func(args AlertFuncArgs) []*Alert {
+		observabilitySiteAlertsDisabled := (args.ViewerFinalSettings != nil && args.ViewerFinalSettings.AlertsHideObservabilitySiteAlerts)
```
Pretty sure you could just call `conf.Get()` here FWIW - but per-user seems nicer anyway because that gives admins the ability to control it for all admins or just themselves

slimsag commented:
> this conf is not accessibly for the conf validator, so the escape hatch for observability.alerts problems is to just remove observability.alerts

SGTM
